### PR TITLE
ARROW-472: [Python] Expose more C++ IO interfaces. Add equals methods to Parquet schemas. Pass Parquet metadata separately in reader

### PR DESCRIPTION
**Pull request by wesm**
Also includes ARROW-471, ARROW-441. 

Needed to compare file schemas easily.

Requires PARQUET-830

### CODE REVIEW

**Comment by wesm** [comment]
@xhochy this is ready to go pending CI passing. Unfortunately ASF JIRA is down

File: cpp/src/arrow/io/file.cc [hunk 1]
```diff
@@ -523,17 +525,24 @@ class MemoryMappedFile::MemoryMappedFileImpl : public OSFile {
...
+    int map_mode;
 
     if (mode != FileMode::READ) {
-      prot_flags |= PROT_WRITE;
-      const bool append = true;
-      RETURN_NOT_OK(OSFile::OpenWriteable(path, append, mode == FileMode::WRITE));
+      // Memory mapping has permission failures if PROT_READ not set
+      prot_flags = PROT_READ | PROT_WRITE;
+      map_mode = MAP_SHARED;
+      static const bool append = true;
```
Comment (xhochy) [inline cpp/src/arrow/io/file.cc]: Why are these variables `static`?

Comment (wesm) [inline cpp/src/arrow/io/file.cc]: Maybe constexpr would be better

**Comment by wesm** [comment]
+1

**Pull request closed without merging**